pkg/controller/remotecluster/rcmgr.go

```diff
@@ -0,0 +1,127 @@
+	"fmt"
+	"time"
+
+	networkingv1 "github.com/oecp/rama/pkg/apis/networking/v1"
+	"github.com/oecp/rama/pkg/rcmanager"
+	"github.com/oecp/rama/pkg/utils"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+func (c *Controller) startRemoteClusterMgr(clusterName string) error {
+	klog.Infof("[debug] processNextRemoteClusterMgr name=%v", clusterName)
+	rcManager, exists := c.rcMgrCache.Get(clusterName)
+	if !exists {
+		klog.Errorf("Can't find rcManager. clusterName=%v", clusterName)
+		return errors.Errorf("Can't find rcManager. clusterName=%v", clusterName)
+	}
+	klog.Infof("Start single remote cluster manager. clusterName=%v", clusterName)
+
+	managerCh := rcManager.StopCh
+	go func() {
+		if ok := cache.WaitForCacheSync(managerCh, rcManager.NodeSynced, rcManager.SubnetSynced, rcManager.IPSynced); !ok {
+			klog.Errorf("failed to wait for remote cluster caches to sync. clusterName=%v", clusterName)
+			return
+		}
+		go wait.Until(rcManager.RunNodeWorker, 1*time.Second, managerCh)
+		go wait.Until(rcManager.RunSubnetWorker, 1*time.Second, managerCh)
+		go wait.Until(rcManager.RunIPInstanceWorker, 1*time.Second, managerCh)
+	}()
+	go rcManager.KubeInformerFactory.Start(managerCh)
+	go rcManager.RamaInformerFactory.Start(managerCh)
+	return nil
+}
+
+// use remove+add instead of update
+func (c *Controller) addOrUpdateRCMgr(rc *networkingv1.RemoteCluster) error {
+	// lock in function range to avoid renewing cluster manager when newing one
+	c.rcMgrCache.mu.Lock()
+	defer c.rcMgrCache.mu.Unlock()
+	klog.Infof("[addOrUpdateRCMgr] cluster=%v", rc.Name)
+
+	clusterName := rc.Name
+	if k, exists := c.rcMgrCache.rcMgrMap[clusterName]; exists {
+		klog.Infof("Delete cluster %v from cache", clusterName)
+		close(k.StopCh)
+		delete(c.rcMgrCache.rcMgrMap, clusterName)
+	}
+
+	rcMgr, err := rcmanager.NewRemoteClusterManager(rc, c.kubeClient, c.ramaClient, c.remoteSubnetLister,
+		c.localClusterSubnetLister, c.remoteVtepLister)
+
+	conditions := make([]networkingv1.ClusterCondition, 0)
+	if err != nil || rcMgr == nil || rcMgr.RamaClient == nil || rcMgr.KubeClient == nil {
+		connErr := errors.Errorf("Can't connect to remote cluster %v", clusterName)
+		c.recorder.Eventf(rc, corev1.EventTypeWarning, "ErrClusterConnectionConfig", connErr.Error())
+		conditions = append(conditions, utils.NewClusterOffline(connErr))
+	} else {
+		conditions = CheckCondition(c, rcMgr.RamaClient, rc.ClusterName, DefaultChecker)
+		rc.Status.UUID = rcMgr.ClusterUUID
+	}
+	rc.Status.Conditions = conditions
+
+	_, err = c.ramaClient.NetworkingV1().RemoteClusters().UpdateStatus(context.TODO(), rc, metav1.UpdateOptions{})
+	if err != nil {
+		runtime.HandleError(err)
+		return err
+	}
+	rcMgr.SetIsReady(IsReady(conditions))
+
+	c.rcMgrCache.rcMgrMap[clusterName] = rcMgr
+	c.rcMgrQueue.Add(clusterName)
```

Comment:
Because I want to decouple the creation and start up of remote cluster manager for the start up may be heavy

Reply:
But it seems quite light now for starting a remote cluster manager, what `rcManager.Run()` do is just start a couple of goroutines and return.

And another benefit what I can think of it that if starting up a remote cluster manager is failed, we can restart it automatically using a work queue. But this will happen only if `rcMgrCache.Get()` is failed, which seems impossible.

Reply:
Hi, bro, also think no need for this, one queue is enough for construction and bootstrap.